feat: docker image tags documentation + tweaks (#26923)
apache/superset: docs/docs/installation/docker.mdx

# Docker Images and Tags

The Apache Superset community extensively uses Docker for development, release,
and productionizing Superset. This page details our Docker builds and tag naming
schemes to help users navigate our offerings.

Images are built and pushed to the [Superset Docker Hub repository](
https://hub.docker.com/r/apache/superset). Different sets of images are created for:
- Published releases, with tags like `3.0.0` and the `latest` tag.
- Pull request iterations, each identified by tags starting with a SHA like
  `8a2f7d378ab13c156fa183d9284b607ed69f5ecc`, and `pr-3454`, referencing the pull
  request ID.
- Merges to the main branch (`master`), resulting in new SHAs, with tags
  prefixed with `master` for the latest `master` version.

Each code version has multiple builds for different purposes, identified by suffixes:

> - **Build Preset:** We offer various images for different needs:

Reviewer: Does "Preset" here refer to the company?  Or does it mean "prefix" or something else?  If it's not referring to Preset the company, it may be clearer to use another term.

Author: Unrelated to the company. I really mean it its it's dictionary definition 
<img width="851" alt="Screenshot 2024-02-01 at 12 06 51 PM" src="https://github.com/apache/superset/assets/487433/da884e0e-d0f6-43e3-9399-41c4956ec3f2">

  - **Lean:** The default Docker image, including both frontend and backend. Tags
    without a build_preset are lean builds, e.g., `latest`.
  - **Dev:** For development, with a headless browser and root access.

>   - **Py310:** Similar to lean but with Python 3.10.

Reviewer: ```suggestion
  - **PyXXX, e.g., Py310:** Similar to lean but with a different Python version (in this example, 3.10).
```

  - **CI:** For certain CI workloads.
  - **WebSocket:** For Superset clusters supporting advanced features.
  - **Dockerize:** Used by Helm.
- **Platform:** We build for `linux/arm64` and `linux/amd64`. The `-arm` suffix
  indicates ARM builds (e.g., `latest-arm`), while tags without a suffix are for

>   AMD (e.g., `latest`).

Reviewer: I see people asking about ARM builds for certain Mac chipsets.  If we know of any that require ARM, maybe we state that here.

Author: Actually when I first got into this, I assumed we introduced `linux/arm64` for Mac workflows, but since then realized there might be a `linux/arm64/v8` for Macs, or at least my M2 laptop seemed to think it needed that. Unclear if Mac M1s use `linux/arm64` or something else. It also brings up the question of if it's not for Macs, how common is `linux/arm64` and should we actually build EVERY COMMIT for EVERY PR on that!?

I personally don't like docker on my dev env much unless I absolutely have to.

Ok, I did a bit of research and while by default my MAC wants that `linux/arm64/v8`, it looks like it'll work with `linux/arm64`, but requires something like this -> 
```
export DOCKER_DEFAULT_PLATFORM=linux/amd64
```

I'll add a note in the docs, though I really don't have much context here. Maybe a v8 build would perform better? If we build on v8, is that compatible with say AWS-hosted arm offering? Does it matter? idk...

Author: @sfirke I added a section bellow "On ARM builds and working with Apple silicon"

Reviewer: Awesome, thanks for doing the legwork and experimentation with Mac & ARM.  People are often asking about it.  I committed two trivial copyediting changes to that section.

## Key Image Tags and Examples

- `latest`: The latest official release build, implicitly the lean build on
  `linux/amd64`.

> - `latest-dev`

Reviewer: ```suggestion
- `latest-dev`: the `-dev` image of the latest official release build, with a headless browser and root access.
```

- `latest-dev`: the `-dev` image of the latest official release build, with a headless browser and root access.
- `master`: The latest build from the `master` branch, implicitly lean on
  `linux/amd64`.
- `master-dev`: Similar to `master` but includes a headless browser and root access.
- `pr-5252`: The latest commit in PR 5252.
- `30948dc401b40982cb7c0dbf6ebbe443b2748c1b-dev-arm`: A `linux/arm64` build for
  this specific SHA, which could be from a pull request, master merge, or release.
- `30948dc-dev-arm`: Same as above, but SHA truncated to 7 characters for a
  shorter handle on the same image
- `websocket-latest`: The WebSocket image for use in a Superset cluster.

For insights or modifications to the build matrix and tagging conventions,
check the [build_docker.py](https://github.com/apache/superset/blob/master/scripts/build_docker.py)
script and the [docker.yml](https://github.com/apache/superset/blob/master/.github/workflows/docker.yml)
GitHub action.

## Caching

To accelerate builds, we follow Docker best practices and use `apache/superset-cache`.

## On ARM builds and working with Apple silicon

Apple's current generation of computers uses ARM-based CPUs, and Docker

> running on MACs seem to require `linux/arm64/v8` (at least my M2 was

Reviewer: ```suggestion
running on MACs seem to require `linux/arm64/v8` (at least one user's M2 was
```

> configured in that way). From my experience, setting the envrionment

Reviewer: ```suggestion
configured in that way). Setting the environment
```

variable `DOCKER_DEFAULT_PLATFORM` to `linux/amd64` seems to function in
term of leveraging, and building upon the Superset builds provided here.

```
export DOCKER_DEFAULT_PLATFORM=linux/amd64
```

Presumably, `linux/arm64/v8` would be more optimized for this generation

> of chips, but less compatible across the ARM ecosystem.

Reviewer 2: This doc is super helpful. 🙌
